### PR TITLE
Kubernetes source plugin

### PR DESCRIPTION
**Pull request by huseyinbabal**
⚠️ Kubernetes plugin extraction was a huge task, but we couldn't figured it out in the beginning. So, you see a failing tests (2 of them only), but I intentionally left them as is since I have a couple of followup tasks for this PR. You can see them all [here](https://github.com/kubeshop/botkube/issues/840)  

## Description

Changes proposed in this pull request:

- Kubernetes plugin is implemented
- Filters are removed from core configuration and  moved to kubernetes plugin
- Actions are managed in core, not in kubernetes source anymore
- Recommendations are part of kubernetes plugin, they are removed from core
- Please see followup tasks [here](https://github.com/kubeshop/botkube/issues/840) 

### Build Kubernetes Plugin
Following command will build kubernetes plugin 
```shell
PLUGIN_DOWNLOAD_URL_BASE_PATH=http://localhost:8080/plugin-dist make gen-plugins-index
```

### Run Plugin Server
```shell
npx serve --listen 8080
```

### Run Botkube With the Following Configurations
```yaml

communications:
  'default-group':
    # Settings for Slack
     socketSlack:
       enabled: true
       appToken: "xapp-..."
       botToken: "xoxb-..."
      notification:
        type: "long" # TODO: Change to `short` during testing
      channels:
        'default':
          name: botkube-demo
          bindings:
            sources:
              # - plugin-based
              # - k8s-events
              - k8s-annotated-cm-delete
              - k8s-updates
              # - k8s-pod-create-events
                #              - plugin-based
            executors:
              - kubectl-plugin
        'secondary':
          name: botkube-demo-secondary
          bindings:
            sources:
              # - plugin-based
              - k8s-events
              # - k8s-annotated-cm-delete
              # - k8s-pod-create-events
                #              - plugin-based
            executors:
              - kubectl-plugin

settings:
  clusterName: gke-playground
  configWatcher: true
  lifecycleServer:
    deployment:
      name: botkube
      namespace: botkube
    port: "2113"

configWatcher:
  enabled: false
  initialSyncTimeout: 0

sources:
  'k8s-events':
    displayName: "K8s recommendations"
    'botkube/kubernetes':
      enabled: true
      config:
        kubeConfig: /Users/huseyin/.kube/config
        log:
          level: debug
        recommendations:
          pod:
            noLatestImageTag: true
            labelsSet: true
          ingress:
            backendServiceValid: false
            tlsSecretValid: false
        namespaces:
          include:
            - botkube
        event:
          types:
            - create
            - update
        resources:
          - type: v1/configmaps
  'k8s-annotated-cm-delete':
    displayName: "K8s ConfigMap delete events"
    'botkube/kubernetes':
      enabled: true
      config:
        kubeConfig: /Users/huseyin/.kube/config
        log:
          level: debug
        namespaces:
          include:
            - botkube
        # labels:
        #   test.botkube.io: "true"
        event:
          types:
            - delete
        resources:
          - type: v1/configmaps

  'k8s-pod-create-events':
    'botkube/kubernetes':
      enabled: true
      config:
        log:
          level: debug
        kubeConfig: /Users/huseyin/.kube/config
        namespaces:
          include:
            - botkube
        event:
          types:
            - create
        resources:
          - type: v1/pods

  'k8s-updates':
    displayName: "K8s ConfigMaps updates"
    'botkube/kubernetes':
      enabled: true
      config:
        log:
          level: debug
        kubeConfig: /Users/huseyin/.kube/config
        namespaces:
          include:
            - default
        event:
          types:
            - create
            - update
            - delete
        resources:
          - type: v1/configmaps
            namespaces:
              include:
                - botkube
            event: # overrides top level `event` entry
              types:
                - update
            updateSetting:
              includeDiff: true
              fields:
                - data
executors:
  'kubectl-plugin':
    ## Kubectl executor configuration.
    botkube/kubectl:
      enabled: true
      # namespaces:
      #   # -- List of allowed Kubernetes Namespaces for command execution.
      #   # It can also contain a regex expressions:
      #   #  `- ".*"` - to specify all Namespaces.
      #   include:
      #     - ".*"
      #   # -- List of ignored Kubernetes Namespace.
      #   # It can also contain a regex expressions:
      #   #  `- "test-.*"` - to specify all Namespaces.
      #   exclude: []
      # # -- If true, enables `kubectl` commands execution.
      # enabled: true
      # ## List of allowed `kubectl` commands.
      # commands:
      #   # -- Configures which `kubectl` methods are allowed.
      #   verbs: ["api-resources", "api-versions", "cluster-info", "describe", "explain", "get", "logs", "top"]
      #   # -- Configures which K8s resource are allowed.
      #   resources: ["deployments", "pods", "namespaces", "daemonsets", "statefulsets", "storageclasses", "nodes", "configmaps", "services", "ingresses"]
      # # -- Configures the default Namespace for executing Botkube `kubectl` commands. If not set, uses the 'default'.
      # defaultNamespace: default
      # # -- If true, enables commands execution from configured channel only.
      # restrictAccess: false

actions:
  'describe-created-resource':
    # -- If true, enables the action.
    enabled: false
    # -- Action display name posted in the channels bound to the same source bindings.
    displayName: "Describe created resource"
    # -- Command to execute when the action is triggered. You can use Go template (https://pkg.go.dev/text/template) together with all helper functions defined by Slim-Sprig library (https://go-task.github.io/slim-sprig).
    # You can use the `{{ .Event }}` variable, which contains the event object that triggered the action. See all available event properties on https://github.com/kubeshop/botkube/blob/main/pkg/event/event.go.
    # @default -- See the `values.yaml` file for the command in the Go template form.
    command: "kubectl describe {{ .Event.Kind | lower }}{{ if .Event.Namespace }} -n {{ .Event.Namespace }}{{ end }} {{ .Event.Name }}"

    # -- Bindings for a given action.
    bindings:
      # -- Event sources that trigger a given action.
      sources:
        #- plugin-based
      # -- Executors configuration used to execute a configured command.
      executors:
        - kubectl-plugin
plugins:
  repositories:
    botkube:
      url: http://localhost:8080/plugins-index.yaml

### To See Notifications in Action

```shell
kubectl run s9 --image=nginx:latest -n botkube
```

```shell
kubectl create configmap my-config-def --from-literal=key1=config1 -n botkube
```

and then update configmap to see diff is shown in slack

## Related issue(s)
Relates #840

### CODE REVIEW

File: pkg/api/executor/grpc_adapter.go
Comment (mszostok) [inline pkg/api/executor/grpc_adapter.go:56]: please revert
```suggestion
		//	  	Data: data,
```

File: pkg/api/source/grpc_adapter.go
Comment (mszostok) [inline pkg/api/source/grpc_adapter.go:223]:
```suggestion
				return nil // output closed, no more messages
```

File: pkg/event/event.go
Comment (mszostok) [inline pkg/event/event.go:74]: I know that we will fix that in the follow-up task. But let's remove that already so we will have a better diff with `internal/source/kubernetes/event/event.go` - in the rest of the code you can import that moved event instead.

File: pkg/k8sutil/resource.go
Comment (mszostok) [inline pkg/k8sutil/resource.go:14]: please remove that file and use `internal/source/kubernetes/k8sutil/resource.go` instead.

**Review by mszostok — COMMENTED** [review]
I still need to check 5 more files:
- internal/source/dispatcher.go
- internal/source/kubernetes/registration.go
- internal/source/kubernetes/router.go
- internal/source/kubernetes/source.go
- internal/source/scheduler.go

And run some manual tests.

**Comment by mszostok** [comment]
You need to move `filters` under `k8s` source:
https://github.com/kubeshop/botkube/blob/09e4a265cc182ee259638bf2daf13e1aac016c7a/helm/botkube/values.yaml#L411-L420

**Comment by mszostok** [comment]
As a part of this PR, please also document all breaking under: https://github.com/kubeshop/botkube/issues/972. Thanks! 🙇

File: internal/source/kubernetes/source.go
Comment (mszostok) [inline internal/source/kubernetes/source.go:474]: probably copy-past error :)

File: internal/source/kubernetes/config/config.go
Comment (pkosiec) [reply internal/source/kubernetes/config/config.go:195]: Yes, I think we can remove `botkube.io/disable`, but I know at least a few people use it, so we should introduce `exclude` annotations and labels instead as a follow-up 🤔

File: cmd/executor/gh/main.go
Comment (pkosiec) [reply cmd/executor/gh/main.go:108]: Actually it is upgrading (but there's 1.23.0 already)

Comment (huseyinbabal) [reply cmd/executor/gh/main.go:108]: I will revert and test it again, I had a problem with gh plugin generation at that time, I will test with up to date version

**Review by pkosiec — COMMENTED** [review]
Great work! Looks like the basic functionality works OK, I tested a few configurations and they worked well. Telemetry also OK 👍 

I agree with all @mszostok comments, and here are just a few ones from my side 🙂

File: internal/source/kubernetes/k8sutil/resource.go
Comment (huseyinbabal) [reply internal/source/kubernetes/k8sutil/resource.go:1]: Let me explain;

As you already know, we haven't implemented all the platforms yet, platform renderers depends on `event` package, and `event` depends on k8sutil, that's why I keep them, we can remove that after platform migration, wdyt?

File: pkg/event/event.go
Comment (huseyinbabal) [reply pkg/event/event.go:74]: Isn't that dangerous to depend a public package to a plugin internals? I just wanted to wait until platform is removed, this event package is still used in a couple of places.

File: pkg/k8sutil/resource.go
Comment (huseyinbabal) [reply pkg/k8sutil/resource.go:14]: As described above, I didn't wanted to depend our core to a plugin internal. I will be removing k8sutils anyways after platform migration, wdyt?

File: internal/source/kubernetes/source.go
Comment (huseyinbabal) [reply internal/source/kubernetes/source.go:380]: I tested other plugins' json specs and it gives schema not found on validation, can we elaborate on that part?

File: internal/source/kubernetes/source.go
Comment (huseyinbabal) [reply internal/source/kubernetes/source.go:438]: showing possible values will prevent flakiness, and I thought that, time to time we can change that options wdyt? I put current values we have in values.yaml, but I can definitely extend that to all possible values

File: internal/source/kubernetes/config/config.go
Comment (huseyinbabal) [reply internal/source/kubernetes/config/config.go:195]: I completely moved the logic, and didn't wanted to touch internals. Handling as followup is fine by me.

File: internal/source/kubernetes/config/config.go
Comment (huseyinbabal) [reply internal/source/kubernetes/config/config.go:18]: ok, I will remove them from config and pass the existing values from our configuration.

File: internal/source/kubernetes/event/event.go
Comment (huseyinbabal) [reply internal/source/kubernetes/event/event.go:32]: There is one usage in node annotation checker, is it ok to keep it for now?

File: cmd/source/kubernetes/main.go
Comment (huseyinbabal) [reply cmd/source/kubernetes/main.go:1]: I just saw this, never happened to me but this is something we need to spot if we see that second time.

File: cmd/source/kubernetes/main.go
Comment (huseyinbabal) [reply cmd/source/kubernetes/main.go:2]: That is because slack renderer has very special functionality to render kubernetes metadata like that, so rendering used to be handled on slack_renderer. However, in plugin structure, we render messages in advance, and I tried to unified them as a general message format. Again, I am open to suggestions here :)

File: pkg/event/event.go
Comment (mszostok) [reply pkg/event/event.go:74]: it's not, actually it's a proper solution to import some parts from `internal` in `pkg`. If you want I can explain on 1:1 call

File: internal/source/kubernetes/config/config.go
Comment (huseyinbabal) [reply internal/source/kubernetes/config/config.go:24]: @pkosiec we don't enable plugins if they are not bound to channels. @mszostok if you are ok too, I will change that logic in a way that plugins will be always enabled.

File: cmd/source/kubernetes/main.go
Comment (huseyinbabal) [reply cmd/source/kubernetes/main.go:3]: @pkosiec we don't enable plugins if they are not bound to channels. @mszostok if you are ok too, I will change that logic in a way that plugins will be always enabled.

File: internal/source/kubernetes/config/config.go
Comment (huseyinbabal) [reply internal/source/kubernetes/config/config.go:24]: Basically, we have command in the context of actions only, that is why I have added like that.

File: internal/source/kubernetes/commander/guard.go
Comment (huseyinbabal) [reply internal/source/kubernetes/commander/guard.go:1]: followup task : extract to pkg to share between kubectl and k8s

File: cmd/source/kubernetes/main.go
Comment (huseyinbabal) [reply cmd/source/kubernetes/main.go:2]: followup

File: cmd/source/kubernetes/main.go
Comment (huseyinbabal) [reply cmd/source/kubernetes/main.go:3]: followup

File: internal/source/kubernetes/config/config.go
Comment (mszostok) [reply internal/source/kubernetes/config/config.go:193]: OK, to do that in the follow-up 👍

File: internal/source/scheduler.go
Comment (mszostok) [inline internal/source/scheduler.go:113]: Please update the comment to reflect current approach. Thanks!

File: pkg/api/source/grpc_adapter.go
Comment (mszostok) [inline pkg/api/source/grpc_adapter.go:27]: Right now there is no option that we will get a list of configs as we start multi-processes. When we will change do documentation and simplify this logic here?

Should we also do that for `executors`? 
Maybe we should do that as a follow-up after RBAC? WDYT @pkosiec @huseyinbabal?

File: internal/source/kubernetes/source.go
Comment (mszostok) [inline internal/source/kubernetes/source.go:91]: Let's clean-up the code in this function.
```go
func (*Source) Stream(ctx context.Context, input source.StreamInput) (source.StreamOutput, error) {
	cfg, err := config.MergeConfigs(input.Configs)
	if err != nil {
		return source.StreamOutput{}, fmt.Errorf("while merging input configs: %w", err)
	}

	s := Source{
		startTime: time.Now(),
		messageCh: make(chan source.Message),
		config:    cfg,
		logger: loggerx.New(loggerx.Config{
			Level: cfg.Log.Level,
		}),
		ctx:         ctx,
		clusterName: input.Context.ClusterName,
		kubeConfig:  input.Context.KubeConfig,
	}

	go consumeEvents(s)

	return source.StreamOutput{
		Message: s.messageCh,
	}, nil
}
```

**Review by mszostok — COMMENTED** [review]
Overall good work 🚀

Once all comments will be resolved I will play with it locally as Paweł did.
Please also track all work that you would like to do in the follow-ups. Thanks!

File: helm/botkube/values.yaml
Comment (mszostok) [inline helm/botkube/values.yaml:135]: please update line 118 from `kubernetes` to `botkube/kubernetes`

File: internal/source/kubernetes/source.go
Comment (mszostok) [inline internal/source/kubernetes/source.go:218]:
```suggestion
		s.logger.Errorf("while running recommendations: %w", err)
		return
```

File: pkg/api/source/grpc_adapter.go
Comment (mszostok) [inline pkg/api/source/grpc_adapter.go:72]:
```go
	StreamOutput struct {
		Event chan Event
	}

	Event struct {
		Message          api.Message
		Metadata         any
		AnonymizedFields map[string]interface{}
	}
```

WDYT about such terminology? @huseyinbabal @pkosiec ?

File: internal/source/kubernetes/commander/guard.go
Comment (huseyinbabal) [reply internal/source/kubernetes/commander/guard.go:1]: I handled that in this PR

File: pkg/api/source/grpc_adapter.go
Comment (huseyinbabal) [reply pkg/api/source/grpc_adapter.go:27]: I can create a followup task for this, on my list. I also want to unify that, didn't wanted to confuse by having single config in sources and multiple configs in executors. Noted followup task.

File: internal/source/kubernetes/source.go
Comment (huseyinbabal) [reply internal/source/kubernetes/source.go:219]: good catch

File: pkg/api/source/grpc_adapter.go
Comment (huseyinbabal) [reply pkg/api/source/grpc_adapter.go:72]: I am ok with the wording, changing accordingly

File: cmd/source/kubernetes/main.go
Comment (huseyinbabal) [reply cmd/source/kubernetes/main.go:2]: @mszostok is on it

File: cmd/source/kubernetes/main.go
Comment (huseyinbabal) [reply cmd/source/kubernetes/main.go:3]: https://github.com/kubeshop/botkube/issues/1000

File: pkg/api/source/grpc_adapter.go
Comment (huseyinbabal) [reply pkg/api/source/grpc_adapter.go:27]: Followup https://github.com/kubeshop/botkube/issues/1001

File: internal/source/kubernetes/config/config.go
Comment (huseyinbabal) [reply internal/source/kubernetes/config/config.go]: changed like suggested

**Comment by huseyinbabal** [comment]
LGTM with minor comments 🚀 

> Overall good work 🚀
> 
> Once all comments will be resolved I will play with it locally as Paweł did. Please also track all work that you would like to do in the follow-ups. Thanks!

I addressed all the feedbacks.

File: helm/botkube/values.yaml
Comment (mszostok) [inline helm/botkube/values.yaml:173]: NOTE: after this merging this PR we won't merge configuration anymore. As a result, enabling both `k8s-err-events` and 'k8s-all-events' results in duplicated `error` messages. Probably will be good to mention that in the release notes/docs.

File: cmd/botkube/main.go
Comment (mszostok) [inline cmd/botkube/main.go:131]: it will be good to get rid of not used clients

File: internal/source/kubernetes/source.go
Comment (mszostok) [inline internal/source/kubernetes/source.go:104]: context.Context always should be as a first parameter

File: pkg/execute/factory.go
Comment (mszostok) [inline pkg/execute/factory.go:10]:
```suggestion
	guard "github.com/kubeshop/botkube/internal/command"
```

let's rename that to make it more readable

File: pkg/filterengine/filterengine.go
Comment (mszostok) [inline pkg/filterengine/filterengine.go:10]: actually this file can be deleted, and in the test we should use `filters.DisableAnnotation` from k8s source. I will handle that on my PR, don't worry about that.

**Review by mszostok — APPROVED** [review]
LGTM 👍 but before merge please address my last comments, and this one too: https://github.com/kubeshop/botkube/pull/985#discussion_r1121533684

Also please wait for @pkosiec feedback and approve. Thanks!

**Review by pkosiec — APPROVED** [review]
🚀 Awesome work!

File: helm/botkube/values.yaml
Comment (huseyinbabal) [reply helm/botkube/values.yaml:173]: Ok, I will be stating that in both release docs and main doc

**Comment by huseyinbabal** [comment]
Addressed all the feedbacks, will check locally and focus on docs changes.

**Pull request merged**
